esx_service/cli/vmdkops_admin.py

```diff
@@ -1006,15 +1006,46 @@ def tenant_access_add(args):
         error_info = "Tenant {0} does not exist".format(args.name)
         return operation_fail(error_info)
 
+    privileges = [d for d in tenant.privileges if d[auth_data_const.COL_DATASTORE] == args.datastore]
+    
+    if privileges:
+        error_info = "Privilege already exists for ({0}, {1})".format(args.name, args.datastore)
+        return operation_fail(error_info)
+
     privileges = generate_privileges(args)
     print privileges
+
+    error_info = check_tenant_access_add_args(privileges)
+    if error_info:
+        return operation_fail(error_info)
+
     error_info = tenant.set_datastore_access_privileges(_auth_mgr.conn, [privileges])
       
     if error_info:
         return operation_fail(error_info)
     else:
         print "tenant access add succeeded"
 
+def check_tenant_access_add_args(privileges):
```

Comment:
This validation should be part of API rather than just Admin CLI so inputs coming via API/UI can be validated as well.